File: src/main/java/com/sems/payments/infrastructure/stripe/StripePaymentAdapter.java
package com.sems.payments.infrastructure.stripe;

import com.sems.payments.domain.services.PaymentProvider;
import com.sems.shared.errors.AppException;
import com.stripe.Stripe;
import com.stripe.exception.SignatureVerificationException;
import com.stripe.exception.StripeException;
import com.stripe.model.Event;
import com.stripe.model.PaymentIntent;
import com.stripe.model.StripeObject;
import com.stripe.model.checkout.Session;
import com.stripe.net.Webhook;
import com.stripe.param.PaymentIntentCreateParams;
import com.stripe.param.checkout.SessionCreateParams;
import jakarta.annotation.PostConstruct;
import java.util.Map;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

/**
 * Adaptador de Stripe.
 *
 * <p>Es el unico punto del sistema que conoce el SDK de Stripe. Traduce entre
 * sus tipos y los del puerto {@link PaymentProvider}, de modo que ni el dominio
 * ni la capa de aplicacion dependen del proveedor.
 */
@Slf4j
@Component
public class StripePaymentAdapter implements PaymentProvider {

    @Value("${stripe.secret-key:}")
    private String secretKey;

    @Value("${stripe.webhook-secret:}")
    private String webhookSecret;

    @Value("${stripe.currency:pen}")
    private String defaultCurrency;

    @PostConstruct
    void init() {
        if (secretKey == null || secretKey.isBlank()) {
            log.warn("STRIPE_SECRET_KEY no esta configurada; los cobros fallaran");
            return;
        }
        Stripe.apiKey = secretKey;
    }

    @Override
    public PaymentMethodDetails getPaymentMethodDetails(String stripePaymentMethodId) {
        try {
            com.stripe.model.PaymentMethod method =
                    com.stripe.model.PaymentMethod.retrieve(stripePaymentMethodId);
            var card = method.getCard();
            if (card == null) {
                return new PaymentMethodDetails(method.getType(), "", "", 0, 0);
            }
            return new PaymentMethodDetails(method.getType(), card.getBrand(), card.getLast4(),
                    card.getExpMonth() == null ? 0 : card.getExpMonth().intValue(),
                    card.getExpYear() == null ? 0 : card.getExpYear().intValue());
        } catch (StripeException e) {
            throw providerFailure("no se pudo leer el medio de pago", e);
        }
    }

    @Override
    public PaymentIntentResult createPaymentIntent(CreatePaymentIntentRequest request) {
        try {
            PaymentIntentCreateParams params = PaymentIntentCreateParams.builder()
                    .setAmount(Math.round(request.amount() * 100))
                    .setCurrency(currencyOrDefault(request.currency()))
                    .setPaymentMethod(request.stripePaymentMethodId())
                    .setConfirm(true)
                    // Sin esto, una tarjeta que pide autenticacion deja el cobro
                    // colgado esperando una redireccion que nadie va a completar.
                    .setAutomaticPaymentMethods(
                            PaymentIntentCreateParams.AutomaticPaymentMethods.builder()
                                    .setEnabled(true)
                                    .setAllowRedirects(
                                            PaymentIntentCreateParams.AutomaticPaymentMethods
                                                    .AllowRedirects.NEVER)
                                    .build())
                    .putMetadata("user_id", nullSafe(request.userId()))
                    .putMetadata("subscription_id", nullSafe(request.subscriptionId()))
                    .putMetadata("payment_id", nullSafe(request.paymentId()))
                    .build();

            PaymentIntent intent = PaymentIntent.create(params);
            return new PaymentIntentResult(intent.getId(), intent.getStatus());
        } catch (StripeException e) {
            throw providerFailure("no se pudo iniciar el cobro", e);
        }
    }

    @Override
    public PaymentIntentResult confirmPaymentIntent(String paymentIntentId) {
        try {
            PaymentIntent intent = PaymentIntent.retrieve(paymentIntentId);
            return new PaymentIntentResult(intent.getId(), intent.getStatus());
        } catch (StripeException e) {
            throw providerFailure("no se pudo confirmar el cobro", e);
        }
    }

    /**
     * Crea una sesion de Stripe Checkout.
     *
     * <p>El usuario paga en una pagina alojada por Stripe, asi que los datos de
     * la tarjeta nunca pasan por nuestros servidores. Los identificadores viajan
     * como metadatos para poder correlacionar el webhook con nuestro registro.
     */
    @Override
    public CheckoutSessionResult createCheckoutSession(CreateCheckoutSessionRequest request) {
        try {
            SessionCreateParams params = SessionCreateParams.builder()
                    .setMode(SessionCreateParams.Mode.PAYMENT)
                    .setSuccessUrl(request.successUrl())
                    .setCancelUrl(request.cancelUrl())
                    .addLineItem(SessionCreateParams.LineItem.builder()
                            .setQuantity(1L)
                            .setPriceData(SessionCreateParams.LineItem.PriceData.builder()
                                    .setCurrency(currencyOrDefault(request.currency()))
                                    .setUnitAmount(Math.round(request.amount() * 100))
                                    .setProductData(SessionCreateParams.LineItem.PriceData
                                            .ProductData.builder()
                                            .setName(request.planName() == null
                                                    ? "Suscripcion SEMS" : request.planName())
                                            .build())
                                    .build())
                            .build())
                    .putMetadata("user_id", nullSafe(request.userId()))
                    .putMetadata("subscription_id", nullSafe(request.subscriptionId()))
                    .build();

            Session session = Session.create(params);
            return new CheckoutSessionResult(session.getId(), session.getUrl());
        } catch (StripeException e) {
            throw providerFailure("no se pudo crear la sesion de pago", e);
        }
    }

    @Override
    public ProviderWebhookEvent parseWebhookEvent(String payload, String signature) {
        if (webhookSecret == null || webhookSecret.isBlank()) {
            throw AppException.internal("stripe webhook is not configured");
        }
        Event event;
        try {
            event = Webhook.constructEvent(payload, signature, webhookSecret);
        } catch (SignatureVerificationException e) {
            // Firma invalida: alguien esta enviando eventos que no vienen de Stripe.
            throw AppException.validation("invalid stripe signature");
        }

        StripeObject object = event.getDataObjectDeserializer().getObject().orElse(null);

        if (object instanceof Session session) {
            Map<String, String> metadata = session.getMetadata() == null
                    ? Map.of() : session.getMetadata();
            double amount = session.getAmountTotal() == null ? 0.0
                    : session.getAmountTotal() / 100.0;
            return new ProviderWebhookEvent(event.getId(), event.getType(), payload,
                    session.getPaymentIntent(), "succeeded",
                    session.getId(), metadata.get("user_id"), metadata.get("subscription_id"),
                    amount, session.getCurrency());
        }

        if (object instanceof PaymentIntent intent) {
            return new ProviderWebhookEvent(event.getId(), event.getType(), payload,
                    intent.getId(), intent.getStatus(),
                    null, null, null, null, null);
        }

        return new ProviderWebhookEvent(event.getId(), event.getType(), payload,
                null, null, null, null, null, null, null);
    }

    private String currencyOrDefault(String currency) {
        return currency == null || currency.isBlank()
                ? defaultCurrency : currency.trim().toLowerCase();
    }

    private static String nullSafe(String value) {
        return value == null ? "" : value;
    }

    private AppException providerFailure(String message, StripeException e) {
        log.error("Stripe: {} ({})", message, e.getMessage());
        return AppException.internal("external provider error: " + message);
    }
}

File: src/Sems.Api/Modules/Payments/Infrastructure/StripePaymentAdapter.cs
using Sems.Api.Modules.Payments.Domain.Services;
using Sems.Api.Shared.Errors;
using Stripe;
using Stripe.Checkout;

namespace Sems.Api.Modules.Payments.Infrastructure;

/// <summary>
/// Adaptador de Stripe.
///
/// <para>Es el unico punto del sistema que conoce el SDK de Stripe. Traduce
/// entre sus tipos y los del puerto <see cref="IPaymentProvider"/>, de modo que
/// ni el dominio ni la capa de aplicacion dependen del proveedor.</para>
/// </summary>
public sealed class StripePaymentAdapter : IPaymentProvider
{
    private readonly string _secretKey;
    private readonly string _webhookSecret;
    private readonly string _defaultCurrency;
    private readonly ILogger<StripePaymentAdapter> _logger;

    public StripePaymentAdapter(IConfiguration configuration, ILogger<StripePaymentAdapter> logger)
    {
        _logger = logger;
        _secretKey = configuration["Stripe:SecretKey"]
                     ?? Environment.GetEnvironmentVariable("STRIPE_SECRET_KEY") ?? string.Empty;
        _webhookSecret = configuration["Stripe:WebhookSecret"]
                         ?? Environment.GetEnvironmentVariable("STRIPE_WEBHOOK_SECRET") ?? string.Empty;
        _defaultCurrency = configuration["Stripe:Currency"]
                           ?? Environment.GetEnvironmentVariable("STRIPE_CURRENCY") ?? "pen";

        if (string.IsNullOrWhiteSpace(_secretKey))
        {
            _logger.LogWarning("STRIPE_SECRET_KEY no esta configurada; los cobros fallaran");
        }
        else
        {
            StripeConfiguration.ApiKey = _secretKey;
        }
    }

    public async Task<PaymentMethodDetails> GetPaymentMethodDetailsAsync(string stripePaymentMethodId,
        CancellationToken ct = default)
    {
        try
        {
            var service = new PaymentMethodService();
            var method = await service.GetAsync(stripePaymentMethodId, cancellationToken: ct);
            var card = method.Card;

            return card is null
                ? new PaymentMethodDetails(method.Type, string.Empty, string.Empty, 0, 0)
                : new PaymentMethodDetails(method.Type, card.Brand, card.Last4,
                    (int)card.ExpMonth, (int)card.ExpYear);
        }
        catch (StripeException ex)
        {
            throw ProviderFailure("no se pudo leer el medio de pago", ex);
        }
    }

    public async Task<PaymentIntentResult> CreatePaymentIntentAsync(
        CreatePaymentIntentRequest request, CancellationToken ct = default)
    {
        try
        {
            var options = new PaymentIntentCreateOptions
            {
                Amount = (long)Math.Round(request.Amount * 100),
                Currency = CurrencyOrDefault(request.Currency),
                PaymentMethod = request.StripePaymentMethodId,
                Confirm = true,
                // Sin esto, una tarjeta que pide autenticacion deja el cobro
                // colgado esperando una redireccion que nadie va a completar.
                AutomaticPaymentMethods = new PaymentIntentAutomaticPaymentMethodsOptions
                {
                    Enabled = true,
                    AllowRedirects = "never"
                },
                Metadata = new Dictionary<string, string>
                {
                    ["user_id"] = request.UserId ?? string.Empty,
                    ["subscription_id"] = request.SubscriptionId ?? string.Empty,
                    ["payment_id"] = request.PaymentId ?? string.Empty
                }
            };

            var intent = await new PaymentIntentService().CreateAsync(options, cancellationToken: ct);
            return new PaymentIntentResult(intent.Id, intent.Status);
        }
        catch (StripeException ex)
        {
            throw ProviderFailure("no se pudo iniciar el cobro", ex);
        }
    }

    public async Task<PaymentIntentResult> ConfirmPaymentIntentAsync(string paymentIntentId,
        CancellationToken ct = default)
    {
        try
        {
            var intent = await new PaymentIntentService().GetAsync(paymentIntentId, cancellationToken: ct);
            return new PaymentIntentResult(intent.Id, intent.Status);
        }
        catch (StripeException ex)
        {
            throw ProviderFailure("no se pudo confirmar el cobro", ex);
        }
    }

    /// <summary>
    /// Crea una sesion de Stripe Checkout.
    ///
    /// <para>El usuario paga en una pagina alojada por Stripe, asi que los datos
    /// de la tarjeta nunca pasan por nuestros servidores. Los identificadores
    /// viajan como metadatos para poder correlacionar el webhook con nuestro
    /// registro.</para>
    /// </summary>
    public async Task<CheckoutSessionResult> CreateCheckoutSessionAsync(
        CreateCheckoutSessionRequest request, CancellationToken ct = default)
    {
        try
        {
            var options = new SessionCreateOptions
            {
                Mode = "payment",
                SuccessUrl = request.SuccessUrl,
                CancelUrl = request.CancelUrl,
                LineItems = new List<SessionLineItemOptions>
                {
                    new()
                    {
                        Quantity = 1,
                        PriceData = new SessionLineItemPriceDataOptions
                        {
                            Currency = CurrencyOrDefault(request.Currency),
                            UnitAmount = (long)Math.Round(request.Amount * 100),
                            ProductData = new SessionLineItemPriceDataProductDataOptions
                            {
                                Name = string.IsNullOrWhiteSpace(request.PlanName)
                                    ? "Suscripcion SEMS" : request.PlanName
                            }
                        }
                    }
                },
                Metadata = new Dictionary<string, string>
                {
                    ["user_id"] = request.UserId ?? string.Empty,
                    ["subscription_id"] = request.SubscriptionId ?? string.Empty
                }
            };

            var session = await new SessionService().CreateAsync(options, cancellationToken: ct);
            return new CheckoutSessionResult(session.Id, session.Url);
        }
        catch (StripeException ex)
        {
            throw ProviderFailure("no se pudo crear la sesion de pago", ex);
        }
    }

    public ProviderWebhookEvent ParseWebhookEvent(string payload, string signature)
    {
        if (string.IsNullOrWhiteSpace(_webhookSecret))
        {
            throw AppException.Internal("stripe webhook is not configured");
        }

        Event stripeEvent;
        try
        {
            stripeEvent = EventUtility.ConstructEvent(payload, signature, _webhookSecret);
        }
        catch (StripeException)
        {
            // Firma invalida: alguien esta enviando eventos que no vienen de Stripe.
            throw AppException.Validation("invalid stripe signature");
        }

        var data = stripeEvent.Data.Object;

        if (data is Session session)
        {
            var metadata = session.Metadata ?? new Dictionary<string, string>();
            var amount = session.AmountTotal is null ? 0.0 : session.AmountTotal.Value / 100.0;

            return new ProviderWebhookEvent(stripeEvent.Id, stripeEvent.Type, payload,
                session.PaymentIntentId, "succeeded", session.Id,
                metadata.GetValueOrDefault("user_id"),
                metadata.GetValueOrDefault("subscription_id"), amount, session.Currency);
        }

        if (data is PaymentIntent intent)
        {
            return new ProviderWebhookEvent(stripeEvent.Id, stripeEvent.Type, payload,
                intent.Id, intent.Status, null, null, null, null, null);
        }

        return new ProviderWebhookEvent(stripeEvent.Id, stripeEvent.Type, payload,
            null, null, null, null, null, null, null);
    }

    private string CurrencyOrDefault(string? currency) =>
        string.IsNullOrWhiteSpace(currency) ? _defaultCurrency : currency.Trim().ToLowerInvariant();

    private AppException ProviderFailure(string message, StripeException ex)
    {
        _logger.LogError("Stripe: {Message} ({Error})", message, ex.Message);
        return AppException.Internal($"external provider error: {message}");
    }
}
